File: h5-app/src/e2ee/device-lifecycle.ts
import { ApiError } from '@/api/http';
import type { E2eeDeviceProfile } from '@/e2ee/device-profile';
import { e2eeSecureStateStorage } from '@/e2ee/secure-state-storage';
import { e2eeSession, type E2eeCommandResult } from '@/e2ee/session';
import { e2eeIdentityService } from '@/services/e2ee-identity-service';
import {
  e2eeMlsApiService,
  registrationMaterialFromCommand,
  registrationMaterialFromRestoredCommand,
  type E2eeDeviceRegistrationMaterial,
} from '@/services/e2ee-mls-api-service';

interface DeviceStorage {
  read(accountId: string): Promise<Uint8Array | null>;
  write(accountId: string, state: Uint8Array): Promise<void>;
  readDeviceProfile(accountId: string): Promise<E2eeDeviceProfile | null>;
  writeDeviceProfile(accountId: string, profile: E2eeDeviceProfile): Promise<void>;
}

interface SessionCore {
  initializeWithRoot(deviceIdentity: string, rootPublicKey?: Uint8Array): Promise<E2eeCommandResult>;
  publicMaterial(state: Uint8Array): Promise<E2eeCommandResult>;
  generateKeyPackage(state: Uint8Array): Promise<E2eeCommandResult>;
}

interface IdentityApi {
  fetchRootIdentity(userId: string): Promise<{ publicKey: Uint8Array }>;
}

interface MlsApi {
  registerDevice(
    deviceId: string,
    deviceLabel: string,
    material: E2eeDeviceRegistrationMaterial,
  ): Promise<{ status: string }>;
  publishKeyPackage(deviceId: string, keyPackage: Uint8Array): Promise<unknown>;
  publishKeyPackages(deviceId: string, keyPackages: Uint8Array[]): Promise<unknown>;
  fetchKeyPackageInventory(deviceId: string): Promise<{
    available: number;
    maxAvailable: number;
  }>;
  listDevices(): Promise<Array<{ id: string; status: string }>>;
}

export class E2eeDeviceNotReadyError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'E2eeDeviceNotReadyError';
  }
}

const KEY_PACKAGE_LOW_WATERMARK = 10;
const KEY_PACKAGE_TARGET = 40;
const KEY_PACKAGE_BATCH_LIMIT = 20;
const REPLENISH_RETRY_AFTER_MS = 60_000;

export class E2eeDeviceLifecycle {
  private readonly replenishLocks = new Map<string, Promise<{ replenished: number }>>();
  private readonly nextRetryAt = new Map<string, number>();

  constructor(
    private readonly storage: DeviceStorage = e2eeSecureStateStorage,
    private readonly core: SessionCore = e2eeSession,
    private readonly identityApi: IdentityApi = e2eeIdentityService,
    private readonly mlsApi: MlsApi = e2eeMlsApiService,
    private readonly newDeviceId: () => string = () => crypto.randomUUID(),
  ) {}

  async ensureReady(accountId: string, deviceLabel: string) {
    let state = await this.storage.read(accountId);
    let profile = await this.storage.readDeviceProfile(accountId);
    if (state && !profile) {
      throw new E2eeDeviceNotReadyError('E2EE 设备状态不完整，拒绝重新生成身份');
    }
    if (!state && profile?.registered) {
      throw new E2eeDeviceNotReadyError('E2EE 已注册设备状态缺失，拒绝重新生成身份');
    }

    let material: E2eeDeviceRegistrationMaterial;
    if (!state) {
      let rootPublicKey: Uint8Array | undefined;
      try {
        rootPublicKey = (await this.identityApi.fetchRootIdentity(accountId)).publicKey;
      } catch (error) {
        if (!(error instanceof ApiError) || error.status !== 404) throw error;
      }
      profile ??= {
        deviceId: this.newDeviceId(),
        deviceLabel,
        registered: false,
        keyPackagePublished: false,
        lastControlSequences: {},
        lastCommitMessageIds: {},
      };
      await this.storage.writeDeviceProfile(accountId, profile);
      material = registrationMaterialFromCommand(
        await this.core.initializeWithRoot(`${accountId}/${profile.deviceId}`, rootPublicKey),
      );
      state = material.state;
      await this.storage.write(accountId, state);
    } else {
      material = registrationMaterialFromRestoredCommand(await this.core.publicMaterial(state));
    }

    if (!profile) throw new E2eeDeviceNotReadyError('E2EE 设备档案缺失');
    let readyProfile = profile;

    if (!readyProfile.registered) {
      const registered = await this.mlsApi.registerDevice(
        readyProfile.deviceId,
        readyProfile.deviceLabel,
        material,
      );
      const status = String(registered.status ?? '');
      if (status === 'pending_approval') {
        // 非首设备保持 pending，服务端拒绝发布 KeyPackage，等待可信设备批准。
        readyProfile = {
          ...readyProfile,
          registered: true,
          keyPackagePublished: false,
          deviceStatus: 'pending_approval',
        };
        await this.storage.writeDeviceProfile(accountId, readyProfile);
        return { profile: readyProfile, state };
      }
      readyProfile = { ...readyProfile, registered: true, deviceStatus: 'active' };
      await this.storage.writeDeviceProfile(accountId, readyProfile);
    }

    if (readyProfile.deviceStatus === 'pending_approval') {
      // 已批准设备恢复能力：服务端状态变为 active 后继续发布 KeyPackage。
      const devices = await this.mlsApi.listDevices();
      const current = devices.find((device) => device.id === readyProfile.deviceId);
      if (current?.status !== 'active') {
        return { profile: readyProfile, state };
      }
      readyProfile = { ...readyProfile, deviceStatus: 'active' };
      await this.storage.writeDeviceProfile(accountId, readyProfile);
    }

    if (readyProfile.deviceStatus !== 'pending_approval' && !readyProfile.keyPackagePublished) {
      const generated = await this.core.generateKeyPackage(state);
      state = generated.field(0);
      await this.storage.write(accountId, state);
      await this.mlsApi.publishKeyPackage(readyProfile.deviceId, generated.field(1));
      readyProfile = { ...readyProfile, keyPackagePublished: true, deviceStatus: 'active' };
      await this.storage.writeDeviceProfile(accountId, readyProfile);
    }

    return { profile: readyProfile, state };
  }

  /**
   * 按低水位补充本设备 KeyPackage 库存。账号级互斥：并发调用共享同一个
   * Promise；失败后进入退避窗口，不阻塞已建立房间的消息链。
   */
  async topUpKeyPackages(accountId: string): Promise<{ replenished: number }> {
    const existing = this.replenishLocks.get(accountId);
    if (existing) return existing;
    const task = this.doTopUp(accountId).finally(() => {
      this.replenishLocks.delete(accountId);
    });
    this.replenishLocks.set(accountId, task);
    return task;
  }

  private async doTopUp(accountId: string): Promise<{ replenished: number }> {
    const profile = await this.storage.readDeviceProfile(accountId);
    if (!profile?.registered || profile.deviceStatus === 'pending_approval' || !profile.keyPackagePublished) {
      if (profile?.deviceStatus === 'pending_approval') {
        throw new E2eeDeviceNotReadyError('E2EE 设备待批准，批准后才能补充 KeyPackage');
      }
      throw new E2eeDeviceNotReadyError('E2EE 设备未完成初始化，无法补充 KeyPackage');
    }
    const retryAt = this.nextRetryAt.get(accountId);
    if (retryAt !== undefined && Date.now() < retryAt) {
      return { replenished: 0 };
    }

    const inventory = await this.mlsApi.fetchKeyPackageInventory(profile.deviceId);
    if (inventory.available >= KEY_PACKAGE_LOW_WATERMARK) {
      this.nextRetryAt.delete(accountId);
      return { replenished: 0 };
    }

    const needed = Math.min(
      KEY_PACKAGE_TARGET - inventory.available,
      KEY_PACKAGE_BATCH_LIMIT,
    );
    if (needed <= 0) {
      this.nextRetryAt.delete(accountId);
      return { replenished: 0 };
    }

    try {
      let state = await this.storage.read(accountId);
      if (!state) throw new E2eeDeviceNotReadyError('E2EE 设备状态缺失');
      const keyPackages: Uint8Array[] = [];
      for (let index = 0; index < needed; index += 1) {
        const generated = await this.core.generateKeyPackage(state);
        state = generated.field(0);
        keyPackages.push(generated.field(1));
      }
      await this.storage.write(accountId, state);
      const result = await this.mlsApi.publishKeyPackages(profile.deviceId, keyPackages);
      const inserted = typeof result === 'object' && result !== null
        ? Number((result as { inserted?: unknown }).inserted)
        : keyPackages.length;
      this.nextRetryAt.delete(accountId);
      return { replenished: Number.isFinite(inserted) ? inserted : keyPackages.length };
    } catch (error) {
      this.nextRetryAt.set(accountId, Date.now() + REPLENISH_RETRY_AFTER_MS);
      throw error;
    }
  }
}

export const e2eeDeviceLifecycle = new E2eeDeviceLifecycle();

File: app/lib/core/e2ee/device_lifecycle.dart
import 'dart:typed_data';

import 'package:redcode_e2ee_core/redcode_e2ee_session.dart';
import 'package:uuid/uuid.dart';

import 'device_profile.dart';
import 'identity_service.dart';
import 'mls_api_service.dart';
import 'secure_state_storage.dart';

class E2eeDeviceNotReadyException implements Exception {
  const E2eeDeviceNotReadyException(this.message);
  final String message;

  @override
  String toString() => message;
}

class E2eeDeviceContext {
  const E2eeDeviceContext({required this.profile, required this.state});

  final E2eeDeviceProfile profile;
  final Uint8List state;
}

class E2eeDeviceLifecycle {
  E2eeDeviceLifecycle({
    E2eeSecureStateStorage? storage,
    E2eeIdentityService? identityService,
    E2eeMlsApiService? api,
    RedcodeE2eeSession? core,
    String Function()? newDeviceId,
  }) : _storage = storage ?? E2eeSecureStateStorage(),
       _identityService = identityService ?? E2eeIdentityService(),
       _api = api ?? E2eeMlsApiService(),
       _core = core ?? const RedcodeE2eeSession(),
       _newDeviceId = newDeviceId ?? const Uuid().v4;

  final E2eeSecureStateStorage _storage;
  final E2eeIdentityService _identityService;
  final E2eeMlsApiService _api;
  final RedcodeE2eeSession _core;
  final String Function() _newDeviceId;

  Future<E2eeDeviceContext> ensureReady({
    required String accountId,
    required String deviceLabel,
  }) async {
    var state = await _storage.read(accountId);
    var profile = await _storage.readDeviceProfile(accountId);
    if (state != null && profile == null) {
      throw const E2eeDeviceNotReadyException('E2EE 设备状态不完整，拒绝重新生成身份');
    }
    if (state == null && profile?.registered == true) {
      throw const E2eeDeviceNotReadyException('E2EE 已注册设备状态缺失，拒绝重新生成身份');
    }

    E2eeDeviceRegistrationMaterial material;
    if (state == null) {
      Uint8List? rootPublicKey;
      try {
        rootPublicKey = (await _identityService.fetchRootIdentity(
          accountId,
        )).publicKey;
      } on E2eeIdentityServiceException catch (error) {
        if (error.statusCode != 404) rethrow;
      }
      profile ??= E2eeDeviceProfile(
        deviceId: _newDeviceId(),
        deviceLabel: deviceLabel,
        registered: false,
        keyPackagePublished: false,
        lastCommitMessageIds: const {},
      );
      await _storage.writeDeviceProfile(accountId, profile);
      final initialized = _core.initialize(
        '$accountId/${profile.deviceId}',
        rootPublicKey: rootPublicKey,
      );
      material = E2eeDeviceRegistrationMaterial.fromCommand(initialized);
      state = material.state;
      await _storage.write(accountId, state);
    } else {
      material = _materialFromRestoredState(state);
    }

    if (!profile!.registered) {
      await _api.registerDevice(
        deviceId: profile.deviceId,
        deviceLabel: profile.deviceLabel,
        material: material,
      );
      profile = profile.copyWith(registered: true);
      await _storage.writeDeviceProfile(accountId, profile);
    }

    if (!profile.keyPackagePublished) {
      final generated = _core.execute(E2eeCommandOperation.generateKeyPackage, [
        state,
      ]);
      state = generated.field(0);
      await _storage.write(accountId, state);
      await _api.publishKeyPackage(
        deviceId: profile.deviceId,
        keyPackage: generated.field(1),
      );
      profile = profile.copyWith(keyPackagePublished: true);
      await _storage.writeDeviceProfile(accountId, profile);
    }

    return E2eeDeviceContext(profile: profile, state: state);
  }

  E2eeDeviceRegistrationMaterial _materialFromRestoredState(Uint8List state) {
    final restored = _core.publicMaterial(state);
    if (restored.fields.length != 6) {
      throw const E2eeCommandException('E2EE 公开材料响应字段数量无效');
    }
    return E2eeDeviceRegistrationMaterial(
      state: restored.field(0),
      keyPackage: Uint8List(0),
      rootPublicKey: restored.field(1),
      rootFingerprint: restored.field(2),
      credential: restored.field(3),
      credentialFingerprint: restored.field(4),
      approvalPublicKey: restored.field(5),
    );
  }
}
